File: CocoaSplit/CSAnimationRunner/CSAnimation.py
from uuid import uuid4
from Quartz import CATransaction,CAMediaTimingFunction
from Foundation import NSLog



class CSAnimation:
    def __init__(self, target, keyPath, animation, **kwargs):
        self.target = target
        self.keyPath = keyPath
        self.animation = animation
        self.isWaitMark = False
        self.isWaitOnly = False
        self.ignore_wait = False
        self.extra_model = None
        self.duration = 0.0
        self.cs_input = None
        self.label = None
        self.end_time = 0
        self.begin_time = 0
        self.completion_handler = None
        self.internal_completion_handler = None
        self.toValue = None
        
        if animation:
            animation.setRemovedOnCompletion_(False)
            animation.setFillMode_("forwards")
            self.duration = animation.duration()
     


        if 'repeatcount' in kwargs:
            self.repeatcount(kwargs['repeatcount'])
        if 'autoreverse' in kwargs:
            self.autoreverse()

        if 'timing' in kwargs:
            self.timingFunction(kwargs['timing'])
        if 'repeatduration' in kwargs:
            self.repeatduration(kwargs['repeatduration'])
    
        if 'extra_model' in kwargs:
            self.extra_model = kwargs['extra_model']

        if 'on_complete' in kwargs:
            self.completion_handler = kwargs['on_complete']
        
        if 'label' in kwargs:
            self.label = kwargs['label']
                
        self.internal_completion_handler = self.set_model_value
        


    def completed(self):
        if self.internal_completion_handler:
            self.internal_completion_handler(self)
        if self.completion_handler:
            self.completion_handler(self)

    def apply(self, begin_time):
        self.begin_time = begin_time
        
        
        if self.target and not self.isWaitMark:
            self.animation.setBeginTime_(begin_time)
            self.uukey = "{0}-{1}".format(self.keyPath, uuid4())
            self.target.addAnimation_forKey_(self.animation, self.uukey)
        if not self.ignore_wait:
            self.end_time = begin_time + self.duration
        return self.duration

    def apply_immediate(self):
        if self.target:
            p_value = self.toValue
            CATransaction.begin()
            #CATransaction.setDisableActions_(True)
            self.target.setValue_forKeyPath_(p_value, self.animation.keyPath())
            if self.extra_model:
                self.extra_model.setValue_forKeyPath_(p_value, self.animation.keyPath())
            CATransaction.commit()


    def set_model_value(self, realme=None):

        if self.target:
            p_layer = self.target.presentationLayer()
            p_value = p_layer.valueForKeyPath_(self.animation.keyPath())
            CATransaction.begin()
            CATransaction.setDisableActions_(True)
            self.target.setValue_forKeyPath_(p_value, self.animation.keyPath())
            if self.extra_model:
                self.extra_model.setValue_forKeyPath_(p_value, self.animation.keyPath())
            self.target.removeAnimationForKey_(self.uukey)
            CATransaction.commit()

    def repeatduration(self, duration):
        if self.animation:
            self.animation.setRepeatDuration_(duration)
        self.duration = duration

    def timingFunction(self, fname):
        if self.animation:
            tfunc = CAMediaTimingFunction.functionWithName_(fname)
            if tfunc:
                self.animation.setTimingFunction_(tfunc)

    def repeatcount(self, r_count):
        if self.animation:
            if r_count == 'forever':
                self.ignore_wait = True
                r_count = float(1e50) #I think this is HUGE_VALF
            else:
                self.duration *= r_count
            self.animation.setRepeatCount_(r_count)

    def autoreverse(self):
        if self.animation:
            self.animation.setAutoreverses_(True)
        self.duration *= 2
        return self


class CSAnimationGroup(CSAnimation):

    def addAnimation(self, animation):
        if not hasattr(self, 'animations'):
            self.animations = [animation]
        else:
            self.animations.append(animation)
        self.animation.setAnimations_(self.animations)


    def set_model_value(self):
        if self.target:
            p_layer = self.target.presentationLayer()
            real_animations = self.animation.animations()
            for anim in real_animations:
                kpath = anim.keyPath()
                p_value = p_layer.valueForKeyPath_(kpath)
                self.target.setValue_forKeyPath_(p_value, kpath)
            
            self.target.removeAnimationForKey_(self.uukey)

File: CocoaSplit/CSAnimationRunner/CSAnimation.js


CSAnimation = function(target, keyPath, animation) {
    
    this.target = target;
    this.keyPath = keyPath;
    this.animation = animation;
    this.isWaitMark = false;
    this.isWaitOnly = false;
    this.ignore_wait = false;
    this.extra_model = null;
    this.duration = 0.0;
    this.cs_input = null;
    this.label = null;
    this.end_time = 0;
    this.begin_time = 0;
    this.completion_handler = null;
    this.internal_completion_handler = null;
    this.toValue = null;
    this.baseLayer = null;
    this.layout = null;
    this.uukey = null;
    this.wait_only = null;
    
    
    
    
    var self = this;
    this.completed = function(what) {
        if (self.internal_completion_handler)
        {
            self.internal_completion_handler(self);
        }
        if (self.completion_handler)
        {
            self.completion_handler(self);
        }
    }
    
    this.apply = function(begin_time) {
        this.begin_time = begin_time;
        if (this.target && !this.isWaitMark)
        {
            this.animation.beginTime = begin_time;
            this.uukey = this.keyPath + "-" + generateUUID();
            applyAnimationAsync(this.target, this.animation, this.uukey);
        }
        
        if (!this.ignore_wait)
        {
            this.end_time = begin_time + this.duration;
        }
        
        return this.duration;
    }
    
    this.waitAnimation = function(duration) {
        CSAnimationBlock.currentFrame().waitAnimation(duration, this);
        return this;
    }
    
    this.apply_immediate = function() {
        if (this.target)
        {
            var p_value = this.toValue;
            CATransaction.begin();
            this.target.setValueForKeyPath(p_value, this.animation.keyPath);
            if (this.extra_model)
            {
                this.extra_model.setValueForKeyPath(p_value, this.animation.keyPath);
            }
            CATransaction.commit();
        }
    }
    
    this.set_model_value = function(realme) {
        if (this.target)
        {
            var p_layer = this.target.presentationLayer();
            if (!p_layer)
            {
                return;
            }
            
            var p_value = p_layer.valueForKeyPath(this.animation.keyPath);
            CATransaction.begin();
            CATransaction.setDisableActions(true);
            this.target.setValueForKeyPath(p_value, this.animation.keyPath);
            if (this.extra_model)
            {
                this.extra_model.setValueForKeyPath(p_value, this.animation.keyPath);
            }
            this.target.removeAnimationForKey(this.uukey);
            CATransaction.commit();
            //this.target = null;
            this.animation.delegate = null;
        }
    }
    
    this.internal_completion_handler = this.set_model_value;
    
    this.repeatduration = function(duration) {
        if (this.animation)
        {
            this.animation.repeatDuration = duration;
        }
        this.duration = duration;
    }
    
    this.timingFunction = function(fname) {
        if (this.animation)
        {
            var tfunc = CAMediaTimingFunction.functionWithName(fname);
            if (tfunc)
            {
                this.animation.setTimingFunction(tfunc);
            }
        }
        return this;
    }
    
    this.repeatforever = function() {
        return this.repeatcount('forever');
    }
    
    
    this.repeatcount = function(r_count) {
        if (this.animation)
        {

            if (r_count == 'forever')
            {

                this.ignore_wait = true;
                r_count = FLT_MAX;
            } else {
                this.duration *= r_count;
            }
            this.animation.repeatCount = r_count;
        }
        return this;
    }
    
    this.autoreverse = function() {
        if (this.animation)
        {
            this.animation.autoreverses = true;
        }
        this.duration *= 2;
        return this;
    }
    
    if (animation)
    {
        animation.removedOnCompletion = 0;
        //animation.fillMode = "forwards";
        this.duration = animation.duration;
    }
    
    this.set_extra_model = function(extra_model) {
        this.extra_model = extra_model;
        return this;
    }
    
    this.set_on_complete = function(on_complete) {
        this.completion_handler = on_complete;
        return this;
    }
    
    this.set_label = function(label) {
        this.label = label;
        return this;
    }
    

}

